framework/src/main/java/org/checkerframework/framework/type/DefaultAnnotatedTypeFormatter.java

```diff
@@ -76,6 +81,35 @@ protected DefaultAnnotatedTypeFormatter(FormattingVisitor visitor) {
     this.formattingVisitor = visitor;
   }
 
+  /**
+   * Maps from nondeterministic capture conversion numbers as output by javac to deterministic ones.
+   * This is useful for comparing output across two runs of the Checker Framework.
+   *
+   * <p>This map is necessary for deterministic and informative output. javac might print two
+   * distinct capture-converted variables as "capture#222" if the second is created after the first
+   * is garbage-collected; this javac output is misleading because it looks like the two printed
+   * representations refer to the same variable.
+   *
+   * <p>This map contains type variables that have been formatted. Therefore, the numbers may differ
+   * between Checker Framework runs if the different runs print different values (say, one of them
+   * prints more type variables than the other).
+   */
+  protected static final Map<TypeVariable, Integer> captureConversionIds =
+      new WeakIdentityHashMap<>();
+
+  /** The last deterministic capture conversion ID that was used. */
+  protected static int prevCaptureConversionId = 0;
+
+  /**
+   * Returns a deterministic capture conversion ID for the given javac captured type.
+   *
+   * @param capturedType a type variable, which must be a capture-converted type variable
+   * @return a deterministic capture conversion ID
+   */
+  static int getCaptureConversionId(TypeVariable capturedType) {
+    return captureConversionIds.computeIfAbsent(capturedType, key -> ++prevCaptureConversionId);
```

Comment:
A static cache of java types has caused memory leaks in the past.  So, this code either needs to be behind a flag that is off by default.  Or instead of storing the TypeVariable can you just store the non-deterministic number that javac prints? So `capture#826 of ? extends java.lang.Object` would appear in the the map as `826 -> 1`.

Reply:
Thanks for the explanation.